### ISSUE TITLE
DeepAR and Deep State Space Model

### ISSUE THREAD

**Issue by AIAficionado** [awslabs/gluon-ts#694]
Hi all,

Firstly, thanks for providing this library for experimental purposes! I am familiar with DL/ML models but recently introduced to time series forecasting. 

I have used various techniques in other libraries such as Hierarchial Bayesian Structural Time Series, this specifically allows to share the statistical strength from multiple time series for a global model. I wanted to implement DeepAR and Deep State Space models. 

However, I am stuck on the implementation it via the library gluonts:

Am I supposed to include the other time series as a dynamical categorical feature in the key value for ListDataset module or on target field in GluonTS?

-  Place the multiple time series in target key value as a list in the target field or 
training_data = ListDataset([{"start": df.index[0], **"target": [df.ts0[:69],df.ts1[:69],df.ts2[:69]]**
, freq = "M"}])

-  Place the multiple time series as dynamical feature?
training_data = ListDataset([{"start": df.index[0], "target": df.ts0[:69], freq = "M", **dynamic_feat=[df.ts1[:69],df.ts2[:69]]}**])

-  How can I implement the multiple time series in a Deep State package [model?](https://gluon-ts.mxnet.io/api/gluonts/gluonts.model.deepstate.html)

Technical Questions
- DeepAR - requires the multiple time series to have future?
- DeepStateSpace - does not require the multiple time series to have future time steps right?

**Comment by jaschau** [comment]
Hi @AlAficionado, at least for DeepAR, if you have a multivariate time series with N time series of length T, target should have shape [N,  T]. In training mode, dynamic_feat_real can also have [N, T]. In prediction mode, target should have shape [N, T] whereas dynamic_feat_real should have shaoe [N, T + prediction_length].

**Comment by AIAficionado** [comment]
> Hi @AlAficionado, at least for DeepAR, if you have a multivariate time series with N time series of length T, target should have shape [N, T]. During training time, feat_dynamic_real can also have shape [N, T]. During test time, however, target should have shape [N, T] whereas feat_dynamic_real should have shape [N, T + prediction_length].
> 
> The difference stems from the fact that during test-time, future values of target are assumed to be unknown whereas future values of feat_dynamic_real are assumed to be known.

Thanks for the clarification.

 If you had multiple time series, that require predictions and have some dependencies across the time series that you would like to model and require predictions for all time series, would you label the other time series as list in target because you can't label them as dynamic_feat_real as you don't know future data?

**Comment by jaschau** [comment]
Yes, if you require predictions of all time series then including all of them in target as [ts1,..., tsN] (which, as a numpy array has shape [N, T]) would be the correct approach.

**Comment by wasimahmadpk** [comment]
I was working on similar problem. I tried to predict multiple related time series by training a global model. When I combined all the time series inside target. It gave dimension problem.

Error: "raise GluonTSDataError(
gluonts.core.exception.GluonTSDataError: Array 'target' has bad shape - expected 1 dimensions, got 2."

**Comment by wasimahmadpk** [comment]
@l-cdc I am doing it like this and it works fine now.`

train_dataset = ListDataset(
    data_iter=[
        {
            "start": "2019-01-01 00:00:00",
            "target": [timeseries1[start:train_stop], timeseries2[start:train_stop]],                      
        },
    ],
    freq=freq,
    one_dim_target=False
`)`

estimator = DeepAREstimator(
    'H', prediction_length=prediction_length, trainer=Trainer(epochs=epochs, hybridize=False),
    distr_output=MultivariateGaussianOutput(dim=2),
`
